### PR TITLE
Introduce new async audit layer using AMQP events

### PR DESCRIPTION
**Pull request by mortenoh**
* Introduces spring-jms
* Adds spring-jms configuration
* Allows for @JmsListener, and JmsTemplate

### CODE REVIEW

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/AuditManager.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    public void sendQueue( String queue, Audit audit )
+    {
+        send( new JmsQueue( queue ), audit );
+    }
+
+    public void sendQueue( String queue, Audit audit, Object data )
+    {
+        send( new JmsQueue( queue ), audit, data );
+    }
+}
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/AuditManager.java]: I think that for a client it should not be necessary to know if the audit message is handled by a topic or queue.
The AuditManager should be responsible for self-configuration and post the Audit message to the correct queue or topic.
Therefore, I'd simplify this interface with just an `audit` method which accept an `Audit` object and an option `Object` data. In fact, I'm not even sure the additional Object is really necessary, the Object could simply be a property of `Audit` - and we could add some logic there like `isPrimitive` `isIdentifiable` etc.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/audit/AuditScope.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @author Luciano Fiandesio
+ */
+public enum  AuditScope {
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/audit/AuditScope.java]: New line for {

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/audit/Auditable.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * This annotation signal the system that the object annotated with it can be processed
+ * by the auditing sub-system.
+ *
+ * @author Luciano Fiandesio
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(value = ElementType.TYPE)
+public @interface Auditable {
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/audit/Auditable.java]: New line for {

File: dhis-2/dhis-services/dhis-service-audit-consumer/src/main/java/org/hisp/dhis/audit/AuditConsumer.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @author Luciano Fiandesio
+ */
+public interface AuditConsumer {
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-audit-consumer/src/main/java/org/hisp/dhis/audit/AuditConsumer.java]: New line for {

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/render/DefaultRenderService.java [hunk 5]
```diff
@@ -115,6 +115,7 @@ public String toJsonAsString( Object value )
         }
         catch ( JsonProcessingException ignored )
         {
+            ignored.printStackTrace();
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/render/DefaultRenderService.java]: Use logging, not print.

File: dhis-2/dhis-support/dhis-support-artemis/src/main/test/org/hisp/dhis/artemis/audit/listener/AnnotationUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
+package org.hisp.dhis.artemis.audit.listener;
+
+import static org.junit.Assert.assertTrue;
```
Comment (larshelge) [inline dhis-2/dhis-support/dhis-support-artemis/src/main/test/org/hisp/dhis/artemis/audit/listener/AnnotationUtilsTest.java]: Move imports after license.

**Pull request merged**